### ISSUE TITLE
Custom HTTP clients could be easier

### ISSUE THREAD

**Issue by Flo354** [stripe/stripe-java#946]
Hi there,

A few days ago we implemented our custom HTTP client for our play application. It's great we can now use the stock Play-WS service provided by Play Framework.

But it could be easier. For request with body, we only have access to the byte array. Isn't it possible to get access to the Map of key/values (since it's a multipart/form-data) directly so we don't have to do weird (and useless) conversions to make it compatible with Play Framework?

Thanks,

**Comment by ob-stripe** [comment]
Hi @Flo354. Could you provide a bit more details about your use case?

I'm not opposed to exposing the map of parameters on the `StripeRequest` object if that's what you're requesting, but in most cases you should not be doing your own encoding.

Stripe's API expects requests to be encoded using `application/x-www-form-urlencoded` encoding (`multipart/form-data` is only used for [file creation requests](https://stripe.com/docs/api/files/create?lang=java)). Unfortunately, form encoding is not a very well defined spec. In particular, there's a lot of ambiguity when it comes to encoding nested lists or maps, both of which are heavily used in Stripe's API. We've made some opinionated choices to encode these parameters in an unambiguous manner, but it's unlikely a third-party encoder would produce the same results.

I'm not familiar with Play WS, but I just took a quick look at the README and it looks like you can pass the request body as a `String` with `body()`, so that would be my recommendation.

**Comment by Flo354** [comment]
Thanks for your answer.

I made a mistake, I was talking about `application/x-www-form-urlencoded`.

My point was, when we have `application/x-www-form-urlencoded` since the post data is a String, it would be cool to be able to access the query string instead of the generated bytes.

Because in this case, what happens is basically:
From Stripe SDK: Map => QueryString => byte[] 
From my HttpClient: byte[] => QueryString

We convert a string to a byte array to a string back. We could avoid this double conversion. 

Same occurs with `multipart/form-data`. We do:
From Stripe SDK: Map => ByteArrayOutputStream => byte[]
From m HttpClient: byte[] => ByteString (akka) => InMemoryBodyWritable (Play WS)

We could have done Map => Source<Http.MultipartFormData.Part<Source<ByteString>, ?>, ?> (play WS)

So basically, I think that what I ask for, is to be able to extends the `StripeRequest` class the same way we can already extends `HttpClient`. This way we can go deeper and avoid unnecessary operations/conversions.

And if something like that is done, it would be great to make classes like `HttpContent`, `FormEncoder`... public (currently package-private)

**Comment by ob-stripe** [comment]
Thanks for the additional details! That makes a lot of sense.

Based on your feedback, I'm thinking of making the following changes:
- add a `getContentString()` method to `HttpContent` to return the content as a `String` instead of a `byte[]`
- add a `getParams()` method to `StripeRequest` to return the (non-flattened) map of parameters
- make `FormEncoder` public
- make `KeyValuePair` public
- make `FormEncoder.flattenParamsValue()` public

This way:
- for `application/x-www-form-urlencoded`, you'd be able to avoid the `byte[] -> String` conversion with `request.content().getContentString()`
- for `multipart/form-data`, you'd be able to map the flattened parameters to `Http.MultipartFormData.Part`s with something like this:

    ```java
    List<KeyValuePair<String, Object>> flatParams = FormEncoder.flattenParamsValue(request.getParams(), null);
    List< Http.MultipartFormData.Part<Source<ByteString, ?>> parts = flatParams
        .map(kvp -> {
          if (kvp.getValue() instanceof String) {
            return new Http.MultipartFormData.DataPart(kvp.getKey(), kvp.getValue());
         } else if (kvp.getValue() instanceof File) {
            return new Http.MultipartFormData.FilePart<>(kvp.getKey(), kvp.getValue().getName(), "binary/octet-stream", kvp.getValue());
         }
        })
        .collect(Collectors.toList());
    ```

(Completely untested, but you get the idea.)

Would that work for you?

**Comment by Flo354** [comment]
Yeah, it would be great that way!

No need to extend the `StripeRequest` class and enough flexibility to build
our own parser.

Thanks a lot,


Le jeu. 23 janv. 2020 à 22:28, Olivier Bellone <notifications@github.com> a
écrit :

**Comment by ob-stripe** [comment]
Just released 17.2.0 with the aforementioned changes.
